Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm

```diff
@@ -0,0 +1,356 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "_WKWebExtensionSQLiteStore.h"
+
+#import "CocoaHelpers.h"
+#import "Logging.h"
+#import "_WKWebExtensionSQLiteDatabase.h"
+#import "_WKWebExtensionSQLiteHelpers.h"
+#import "_WKWebExtensionSQLiteRow.h"
+#import <sqlite3.h>
+#import <wtf/BlockPtr.h>
+#import <wtf/RunLoop.h>
+#import <wtf/WeakObjCPtr.h>
+
+using namespace WebKit;
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+@implementation _WKWebExtensionSQLiteStore
+
+- (instancetype)initWithUniqueIdentifier:(NSString *)uniqueIdentifier directory:(NSString *)directory usesInMemoryDatabase:(BOOL)useInMemoryDatabase
+{
+    ASSERT_WITH_MESSAGE(![self.class isEqual:_WKWebExtensionSQLiteStore.class], "Must instantiate a concrete subclass instead.");
+
+    if (!(self = [super init]))
+        return nil;
+
+    _uniqueIdentifier = [uniqueIdentifier copy];
+    _directory = [NSURL fileURLWithPath:directory];
+    _useInMemoryDatabase = useInMemoryDatabase;
+
+    NSString *extensionDatabaseQueueName = [NSString stringWithFormat:@"com.apple.WebKit.WKWebExtensionSQLiteStore.%@", _uniqueIdentifier];
+    _databaseQueue = dispatch_queue_create([extensionDatabaseQueueName cStringUsingEncoding:NSUTF8StringEncoding], DISPATCH_QUEUE_SERIAL);
+
+    return self;
+}
+
+- (void)dealloc
+{
+    _WKWebExtensionSQLiteDatabase *database = _database;
+    dispatch_async(_databaseQueue, ^{
+        [database close];
+    });
+}
+
+- (void)deleteStorageWithCompletionHandler:(void (^)(NSError *error))completionHandler
+{
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0), ^{
+        NSError *error;
+        [NSFileManager.defaultManager removeItemAtURL:self->_directory error:&error];
+        completionHandler(error);
+    });
+}
+
+- (void)deleteDatabaseWithCompletionHandler:(void (^)(NSString *errorMessage))completionHandler
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionSQLiteStore>(self);
```

Comment:
Since this is ObjC++ and ARC, you should just be able to do `__weak auto *weakSelf = self;`. Otherwise `WeakObjCPtr { self }` would be the simple form for `WeakObjCPtr`.

Reply:
`__weak auto *weakSelf = self;` didn't work so i did `auto weakSelf = WeakObjCPtr<_WKWebExtensionSQLiteStore> { self };`

Reply:
`WeakObjCPtr { self }` should work too, the type will be inferred.

Reply:
I got an error when I did it without the type

Reply:
Oh interesting! The other smart pointers work like that, so maybe they have some extra template magic that makes it work.